include/dxc/HLSL/HLMatrixType.h

```diff
@@ -0,0 +1,97 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// HLMatrixType.h                                                            //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "llvm/IR/IRBuilder.h"
+
+namespace llvm {
+  template<typename T>
+  class ArrayRef;
+  class Type;
+  class Value;
+  class Constant;
+  class StructType;
+  class VectorType;
+  class StoreInst;
+}
+
+namespace hlsl {
+
+class DxilFieldAnnotation;
+class DxilTypeSystem;
+
+// A high-level matrix type in LLVM IR.
+//
+// Matrices are represented by an llvm struct type of the following form:
+// { [RowCount x <ColCount x RegReprTy>] }
+// Note that the element type is always in its register representation (ie bools are i1s).
+// This allows preserving the original type and is okay since matrix types are only
+// manipulated in an opaque way, through intrinsics.
+//
+// During matrix lowering, matrices are converted to vectors of the following form:
+// <RowCount*ColCount x Ty>
+// At this point, register vs memory representation starts to matter and we have to
+// imitate the codegen for scalar and vector bools: i1s when in llvm registers,
+// and i32s when in memory (allocas, pointers, or in structs/lists, which are always in memory).
+//
+// This class is designed to resemble a llvm::Type-derived class.
+class HLMatrixType
+{
+public:
+  static constexpr const char* StructNamePrefix = "class.matrix";
```

Comment:
Prefix should end in '.' so it can't match a class name that starts with 'matrix', like `matrixA`.

Reply:
Hmm you're right. I didn't invent this though, it comes from the snippet below. I wasn't able to make this cause issues but it seems likely it could.

```
bool IsHLSLMatrixType(Type *Ty) {
  if (StructType *ST = dyn_cast<StructType>(Ty)) {
    Type *EltTy = ST->getElementType(0);
    if (!ST->getName().startswith("class.matrix"))
      return false;
```